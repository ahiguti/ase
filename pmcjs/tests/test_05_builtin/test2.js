
var core = Perl.Functions['CORE'];

function dumparr(x, mess) {
  if (mess) {
    print_nonl(mess + ": ");
  }
  print_nonl('[');
  for (var i = 0; i < x.length; ++i) {
    print_nonl(' ' + x[i]);
  }
  print(' ]');
}

function dotest_arr(f) {
  var arr = [ 99, 98, 97 ];
  // var arr = Perl.NewArray([ 99, 98, 97 ]);
  f(arr);
  dumparr(arr);
}

dotest_arr(function (arr) { print("pop=" + core.pop(arr)); });
dotest_arr(function (arr) { core.push(arr, 15); });
dotest_arr(function (arr) { print("shift=" + core.shift(arr)); });
dotest_arr(function (arr) { core.unshift(arr, 15); });

dumparr(core.reverse([ 10, 20, 30 ]));
print(core.join(':', [ 10, 20, 30 ]));
dumparr(core.map(function(x) { return x * 2; }, [ 10, 20, 30 ]));

var f1 = function() { return Perl.ARG > 30; }
var x = core.grep(f1, [ 10, 300, 90, 30, 31 ]);
dumparr(x);
var f2 = function(x) { return x > 30; }
var x = core.grep(f2, [ 10, 300, 90, 30, 31 ]);
dumparr(x);

var compare_func = function (x, y) {
  if (x > y) return -1;
  if (x < y) return 1;
  return 0;
};
var arr = [ 'foo', 'bar', 'baz', 'fubar', '0', '99', 'ABCDE' ];
var x = core.sort(compare_func, arr);
dumparr(x);

var s = core.pack('CCCC', [ 65, 66, 67, 68 ]);
print(s);
dumparr(core.unpack('CCCC', s));

function dosplice(f) {
  var arr = [ 99, 98, 97 ];
  // var arr = Perl.NewArray([ 99, 98, 97 ]);
  dumparr(arr, "sporig");
  var x = f(arr);
  dumparr(arr, "spsrc");
  dumparr(x, "spret");
}

dosplice(function(arr) { return core.splice(arr); });
dosplice(function(arr) { return core.splice(arr, 1); });
dosplice(function(arr) { return core.splice(arr, 1, 1); });
dosplice(function(arr) { return core.splice(arr, 1, 1, 999, 998, 997); });

