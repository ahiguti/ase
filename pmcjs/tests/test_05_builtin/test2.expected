pop=97
[ 99 98 ]
[ 99 98 97 15 ]
shift=99
[ 98 97 ]
[ 15 99 98 97 ]
[ 30 20 10 ]
10:20:30
[ 20 40 60 ]
[ 300 90 31 ]
[ 300 90 31 ]
[ fubar foo baz bar ABCDE 99 0 ]
ABCD
[ 65 66 67 68 ]
sporig: [ 99 98 97 ]
spsrc: [ ]
spret: [ 99 98 97 ]
sporig: [ 99 98 97 ]
spsrc: [ 99 ]
spret: [ 98 97 ]
sporig: [ 99 98 97 ]
spsrc: [ 99 97 ]
spret: [ 98 ]
sporig: [ 99 98 97 ]
spsrc: [ 99 999 998 997 97 ]
spret: [ 98 ]
